### PR TITLE
Issue 113: Integration testing for `latent_individual`

### LINKED ISSUE #113: Integration testing for `latent_individual`
## Goal

We would like to test that functions for the `latent_individual` model work well together as a part of a whole pipeline.

## Context

The package is starting to have testing (unit and integration) implemented. These features are important for us to continue to develop the package without introducing bugs.

Originally, I had began implementing integration tests as a part of #107. I am now spliting unit tests and integration tests after a steer during check-in meeting.

## Required features

* Tests of the full `epidist::epidist()` function for the `latent_individual` model
  * Certainly: Stan compiles
  * I think there is another thing here in between "Stan compiles" and convergence (?)
  * Certainly: convergence
  * Possibly: attaining particular values (within some tolerance)
* Equivalent tests for possible settings of the `epidist::epidist` function
  * Certainly: gamma delay
  * Certainly: some modification of the formula
  * Possibly: other model changes e.g. prior

## Related documents

* #107
* #23

### PR DESCRIPTION
**Pull request by athowes**
## Description

This PR will close #113.

I have added integration tests for Stan code compliation and model convergence for the `latent_individual` model with:

* default
* gamma delay
* alternative formula (`1 + sex`)

Remains to do:

* [x] Add samples from the prior and compare to `epidist_prior` (may have to regex the prior to sample from)
* [x] Add test that simulated parameters are recovered up to some tolerance
  * For the gamma one ideally would simulate gamma data (could be a new PR)
  * `sex` should be near zero

Out of scope:

* Making test suite reproducible, have added issue https://github.com/epinowcast/epidist/issues/130
* It's not obvious to me how to recover the gamma settings, so I'm going to make that out of scope for this PR. See issue #131 

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [x] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [x] I have tested my changes locally.
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] My code follows the established coding standards.
- [ ] I have added a news item linked to this PR.
- [x] I have reviewed CI checks for this PR and addressed them as far as I am able.

### CODE REVIEW

**Comment by athowes** [comment]
For the fails I'm getting:

```
══ Failed tests ════════════════════════════════════════════════════════════════
── Failure ('test-int-latent_individual.R:5:3'): epidist.epidist_latent_individual Stan code compiles in the default case ──
Expected `rstan::stan_model(model_code = stancode)` to run without any errors.
i Actually got a <simpleError> with text:
  Boost not found; call install.packages('BH')
── Failure ('test-int-latent_individual.R:28:3'): epidist.epidist_latent_individual Stan code compiles in the gamma delay case ──
Expected `rstan::stan_model(model_code = stancode_gamma)` to run without any errors.
i Actually got a <simpleError> with text:
  Boost not found; call install.packages('BH')
── Failure ('test-int-latent_individual.R:51:3'): epidist.epidist_latent_individual Stan code compiles for an alternative formula ──
Expected `rstan::stan_model(model_code = stancode_sex)` to run without any errors.
i Actually got a <simpleError> with text:
  Boost not found; call install.packages('BH')

[ FAIL 3 | WARN 0 | SKIP 0 | PASS 19 ]
Error: Error: Test failures
```

So something like `Boost not found; call install.packages('BH')`.

This is a bit odd because compiling the Stan code should be easier than running it and achieving convergence. Is there another command that would work better here than `expect_no_error(rstan::stan_model(model_code = stancode))`?

**Comment by athowes** [comment]
* For compile check try using `cmdstanr` and it's `compiles` (need to load model first `cmdstan_model`)

**Comment by athowes** [comment]
* Could `mod <- cmdstan_model(..)` then look if there is valid Stan code before the compile check
* Need to disentangle our code and `brms` code (create new issue for this)

**Comment by athowes** [comment]
I think just remains here to:

* [ ] Figure out how to get the gamma draws out of `fit_gamma`
* [ ] Pass any missing tests that can be feasibly passes

File: tests/testthat/test-int-latent_individual.R [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  expect_no_error(mod_sex$compile())
+})
+
+test_that("epidist.epidist_latent_individual recovers no sex effect when none is simulated", { # nolint: line_length_linter.
+  prep_obs$sex <- rbinom(n = nrow(prep_obs), size = 1, prob = 0.5)
+  formula_sex <- epidist_formula(prep_obs, delay_central = ~ 1 + sex,
+                                 sigma = ~ 1 + sex)
+  fit_sex <- epidist(data = prep_obs, formula = formula_sex)
+  draws <- posterior::as_draws_df(fit_sex$fit)
+  expect_equal(mean(draws$b_sex), 0, tolerance = 0.1)
```
Comment (athowes) [inline tests/testthat/test-int-latent_individual.R:99]: Should be no sex effect for `delay_central` or `sigma` (i.e. two tests)

**Comment by athowes** [comment]
Fails here:

* R-CMD-as-cran-check
  * `Error: CmdStan path has not been set yet. See ?set_cmdstan_path.`
    * Solution here to skip these tests on CRAN? How to get `cmdstan` to be available? I don't know
  * Warning: '::' or ':::' import not declared from: ‘rstan’

**Comment by athowes** [comment]
They all passed! I think this should be good to merge @seabbs but welcome any review

File: DESCRIPTION [hunk 2]
```diff
@@ -42,7 +42,8 @@ Suggests:
     gt,
     dplyr,
     knitr,
-    roxyglobals
+    roxyglobals,
+    rstan
```
Comment (seabbs) [inline DESCRIPTION:46]: I ideally we want to move away from needing to use `rstan` and instead depend on `posteriors` for anything we can't get from `brms`

File: tests/testthat/test-int-latent_individual.R [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  prior_samples <- epidist(data = prep_obs, fn = brms::brm,
+                           sample_prior = "only")
+  lognormal_draws <- extract_lognormal_draws(prior_samples)
+  prior <- epidist_prior(data = prep_obs)
+  param1 <- extract_normal_parameters_brms(prior[1, ])
+  param2 <- extract_normal_parameters_brms(prior[2, ])
+  samples1 <- rnorm(1000, mean = param1$mean, sd = param1$sd)
+  samples2 <- exp(rnorm(1000, mean = param2$mean, sd = param2$sd))
+  # suppressWarnings here used to prevent warnings about ties
+  ks1 <- suppressWarnings(stats::ks.test(lognormal_draws$meanlog, samples1))
```
Comment (seabbs) [inline tests/testthat/test-int-latent_individual.R:30]: Lovely stuff

**Review by seabbs — APPROVED** [review]
This is great. Only note is we should try and use `posterior` over `rstan` as much as possible. Making a new issue for that

**Pull request merged**